Volumemgr support for kubevirt eve

zedi-pramodh opened the pull request:
Implements the new volumehandler, csihandler to create kubernetes specific volumes Implements the volumehandler interface functions like CreateVolume, DestroyVolume etc. for PersistentVolumes (Kubernetes based volumes) Though longhorn can support filemode volumes we only create blockmode volumes since eve only deals with block volumes.

CreateVolume() does the following:
 1) A blank volume create request, creates a blockmode PVC
 2) A downloaded volume qcow2 is converted to blockmode PVC
 3) For container images, we follow the same path of extracting rootfs to /persist/vault/volumes/*container directory
    once the rootfs is laid out, the files required for runx-initrd are generated and copied into /persist/vault/volumes/*container
    The directory content will look like this:
    4b6da250-6608-4686-a9fd-89e16f39ded6:/persist/vault/volumes/fe328d99-a380-45a7-99af-612b6264f628#0.container# ls
    cmdline            environment        image-config.json  modules            rootfs             ug
    This entire directory is first copied to a qcow2 using copy-image-to-qcow.sh script
    The qcow2 is copied to PVC, which will be used as bootdisk for the VM.
    The qcow2 will be deleted to save space.

Volume snapshots are not supported yet and a corresponding error is returned. Kubenertes naming convention does not support "#", so all PVC are keyed with -pvc- instead of "#" Added docker.io/ prefix to the container images inserted into the containerd repository

OhmSpectator commented on pkg/pillar/cmd/volumemgr/volumemgr.go:
```diff
@@ -237,9 +242,23 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject, ar
+
 	if ctx.persistType == types.PersistZFS {
 		if isZvol, _ := zfs.IsDatasetTypeZvol(types.SealedDataset); isZvol {
+			// This code is called only on kubevirt eve
```
But this code is not under
```
if ctx.hvTypeKube {
```
How are we sure, it's only kubevirt-related?

zedi-pramodh replied:
@andrewd-zededa can answer that. I cross checked with him. Basically it means /persist/vault/ which is made out of zvol. That is only possible in Kubevirt eve case. In regular ZFS installs /persist/vault is a filesystem.

eriknordmark replied:
But shouldn't we check a consistent predicate for that? Sometimes checking hvType and somethings checking the dataset doesn't make the code easier to understand hence might introduce bugs down the road.

zedi-pramodh replied:
Actually I will let @andrewd-zededa comment on that. That code already exists I just added a comment to be clear.

eriknordmark replied:
This question is still open.

zedi-pramodh replied:
That code is nothing to do with this commit. I just added comment for the clarity.  I believe Andrew actually started with the hvType and he was made to change that in the reviews. @andrewd-zededa can you comment on that.

OhmSpectator commented on pkg/pillar/types/volumetypes.go:
```diff
@@ -33,6 +33,11 @@ type VolumeConfig struct {
 func (config VolumeConfig) Key() string {
+	// PVC names should not include # so lets choose -pvc-
+	if base.IsHVTypeKube() {
+		return fmt.Sprintf("%s-pvc-%d", config.VolumeID.String(),
```
I'm not sure, but I guess `#` before the generation can be used somewhere in the code for parsing to split the name... Is it possible to have the name with `#%d` at the end?

OhmSpectator replied:
And in general, do we really need the Key of another form for the kubevirt case?.. I may be wrong, but it may be a case in which some other component has an implicit expectation from the Volume[Ref]Config/Status name...

zedi-pramodh replied:
Kubernetes by standard does not support names with "#"

eriknordmark replied:
Did you change the code which parses the name (to extact the generationcounter)?

eriknordmark replied:
FWIW that code is in
pkg/pillar/cmd/volumemgr/initialvolumestatus.go:		volumeIDAndGeneration = strings.ReplaceAll(volumeIDAndGeneration, "#", ".")

It might be simpler to pick some other unique character (which is not in a UUID nor in a number) over -pvc- so that you can do two reaplaceall e.g., from # and from @ to period in the parser.

Last but not least, you don't need to change the Key() functions at all; those are pubsub keys.
You only need to change the function(s) which form the filename and the above parser of the filename.

And please make sure that the population from the PVC also determines the generationCounter so the VolumeStatus is correct.

eriknordmark replied:
This inconsistency still seems like the source of a future bug as well as confusion in the naming of the volumes.

Can we use % in a PVC name? If so that would make more sense than -pvc-.
And with appropriate comments and also code in pkg/pillar/cmd/volumemgr/initialvolumestatus.go to treat % as an alternative to #, then I think there is less risk of confusion and bugs.

zedi-pramodh replied:
Unfortunately % is also not supported. I tried to create a PVC with name testing-%-0

Resource: "/v1, Resource=persistentvolumeclaims", GroupVersionKind: "/v1, Kind=PersistentVolumeClaim"
Name: "testing-%-0", Namespace: "default"
from server for: "/persist/pvc.yaml": invalid resource name "testing-%-0": [may not contain '%']

zedi-pramodh replied:
Actually thinking about it, why does the name even need to have "#" or "%"
Can't we just have UUID-<Generationcounter> ?

eriknordmark replied:
We could have but changing it now means makes upgrades and fallback hard, because existing devices with existing volumes have "#". We can make cmd/upgradeconverter rename those files in /persist/{vault,clear}/volumes, but what if the upgrade 10 minute test fails and it falls back to the old release? We can't loose the existing volumes in that case.

eriknordmark replied:
Latest code addresses this. Thanks.

christoph-zededa commented on pkg/pillar/volumehandlers/csihandler.go:
```diff
@@ -0,0 +1,331 @@
+	// removing of partially created objects must be done inside caller
+	createContext, createCancel := context.WithCancel(createContext)
+
+	done := make(chan bool, 1)
```
```suggestion
	done := make(chan struct{}, 1)
```

OhmSpectator replied:
@christoph-zededa I'm just curious: why is this approach better?

christoph-zededa replied:
It does not imply to the reader of the code that you have the choice to send true or false and that this makes a difference.

christoph-zededa commented on pkg/pillar/volumehandlers/csihandler.go:
```diff
@@ -0,0 +1,331 @@
+
+	log.Noticef("%s args %v", imageToQcowScript, args)
+
+	output, err := base.Exec(log, imageToQcowScript, args...).WithContext(ctx).WithUnlimitedTimeout(432000 * time.Second).CombinedOutput()
```
```suggestion
	output, err := base.Exec(log, imageToQcowScript, args...).WithContext(ctx).WithUnlimitedTimeout(120 * time.Hour).CombinedOutput()
```
just to make it clearer that it is expected to take up to 5 days

zedi-pramodh replied:
At the same time I could not fix a approx time since if TBs of volume is getting copied it could take multiple hours. Any error during the copy will return immediately. So technically I am not seeing an issue with having arbitrarily long timeout here. May be I will change to ZFS zvol timeouts which is (3600 * time.Second) that code was tested at VW.

zedi-pramodh replied:
Oh sorry ZVOL timeout is (120 * time.Hour), looks like that is what you are suggesting, will change to that.

zedi-pramodh replied:
Done

christoph-zededa commented on pkg/pillar/volumehandlers/csihandler.go:
```diff
@@ -0,0 +1,331 @@
+	output, err := base.Exec(log, imageToQcowScript, args...).WithContext(ctx).WithUnlimitedTimeout(432000 * time.Second).CombinedOutput()
+
+	if err != nil {
+		errStr := fmt.Sprintf("CopyImgDirToRawImg: Failed to create raw image file  %s: %v", output, err)
```
```suggestion
		errStr := fmt.Errorf("CopyImgDirToRawImg: Failed to create raw image file  %s: %v", output, err)
```

OhmSpectator replied:
If we apply this change, the var should be changed as well: `err`, not `errStr` and returned directly without calling `error.New()`...

zedi-pramodh replied:
Done

christoph-zededa commented on pkg/pillar/types/volumetypes.go:
```diff
@@ -137,12 +142,19 @@ type VolumeStatus struct {
 	WWN                     string
 	Target                  zconfig.Target
 	CustomMeta              string
+	IsAppImage              bool // Does this volume contain app image
```
```suggestion
	ContainsAppImage              bool // Does this volume contain app image
```
The comment somehow contradicts the meaning of the variable name.

zedi-pramodh replied:
I think I will change comment than the variable.

zedi-pramodh replied:
Done

christoph-zededa commented on pkg/pillar/cas/containerd.go:
```diff
@@ -705,6 +708,82 @@ func (c *containerdCAS) PrepareContainerRootDir(rootPath, reference, rootBlobSha
+		// Userid and GID are same
+		ug := fmt.Sprintf("%d %d", 0, 0)
+		if user != "" {
+			uid, _ := strconv.Atoi(user)
```
please don't _ errors

zedi-pramodh replied:
ok

zedi-pramodh replied:
Done

christoph-zededa commented:
The PR does not have any tests.

christoph-zededa commented on pkg/pillar/cas/containerd.go:
```diff
@@ -705,6 +708,82 @@ func (c *containerdCAS) PrepareContainerRootDir(rootPath, reference, rootBlobSha
+			keyAndValueSlice := strings.SplitN(e, "=", 2)
+			if len(keyAndValueSlice) == 2 {
+				//handles Key=Value case
+				envContent = envContent + fmt.Sprintf("export %s=\"%s\"\n", keyAndValueSlice[0], keyAndValueSlice[1])
```
@rene this reminds me about your PR (https://github.com/lf-edge/eve/pull/3760/files ) - can we use the same code base here?

christoph-zededa commented on pkg/pillar/cas/containerd.go:
```diff
@@ -705,6 +708,82 @@ func (c *containerdCAS) PrepareContainerRootDir(rootPath, reference, rootBlobSha
+		}
+
+		// Userid and GID are same
+		ug := fmt.Sprintf("%d %d", 0, 0)
```
```suggestion
		ug := "0 0"
```

zedi-pramodh replied:
Is that technically wrong ?

christoph-zededa replied:
It just looks like as if a planned code change was forgotten.

zedi-pramodh replied:
Done

zedi-pramodh commented:
> The PR does not have any tests.

Tests are same as what we already have today for volume creation, EVE API is note changing.

christoph-zededa commented:
> Tests are same as what we already have today for volume creation, EVE API is note changing.

According to https://github.com/lf-edge/eve/pull/3768#issuecomment-1955191138 0% of the code changes are tested.
It doesn't have to bested all the way through the EVE API, other tests count into the coverage statistics as well.

zedi-pramodh commented:
> > Tests are same as what we already have today for volume creation, EVE API is note changing.
> 
> According to [#3768 (comment)](https://github.com/lf-edge/eve/pull/3768#issuecomment-1955191138) 0% of the code changes are tested. It doesn't have to be tested all the way through the EVE API, other tests count into the coverage statistics as well.



> It doesn't have to be tested all the way through the EVE API, other tests count into the coverage statistics as well.

Correct. that is expected for now. This code will be executed only on kubevirt eve. We do not yet test kubevirt eve for code coverage since we cannot even build kubevirt eve yet from master branch. We are in the process of sending PRs for review. Once we have all PRs merged we can start enabling code coverage testing for kubevirt eve

zedi-pramodh commented on pkg/pillar/volumehandlers/csihandler.go:
```diff
@@ -0,0 +1,331 @@
+	done := make(chan bool, 1)
+
+	defer func() {
+		done <- true
```
Done

christoph-zededa commented:
> > > Tests are same as what we already have today for volume creation, EVE API is note changing.
> > 
> > 
> > According to [#3768 (comment)](https://github.com/lf-edge/eve/pull/3768#issuecomment-1955191138) 0% of the code changes are tested. It doesn't have to be tested all the way through the EVE API, other tests count into the coverage statistics as well.
> 
> > It doesn't have to be tested all the way through the EVE API, other tests count into the coverage statistics as well.
> 
> Correct. that is expected for now. This code will be executed only on kubevirt eve. We do not yet test kubevirt eve for code coverage since we cannot even build kubevirt eve yet from master branch. We are in the process of sending PRs for review. Once we have all PRs merged we can start enabling code coverage testing for kubevirt eve

You can for example do it like this: https://github.com/christoph-zededa/eve/commit/25be593db3bd3e22acebe30a38e11038d2e42211

zedi-pramodh commented:
Thanks @christoph-zededa can I actually go ahead and use the code you suggested ?

eriknordmark commented on pkg/pillar/cmd/volumemgr/handlework.go:
```diff
@@ -248,6 +256,14 @@ func casIngestWorker(ctxPtr interface{}, w worker.Work) worker.WorkResult {
+	} else {
+		if ctx.hvTypeKube {
+			cfg := lookupContentTreeConfig(ctx, status.ContentID.String())
+			if cfg != nil && cfg.IsAppImage {
```
Does this imply there might be a race where OciImageName might not be set e.g., due to an error?
What about e.g., the case when the ContentTreeConfig has been deleted by zedagent but volumemgr has not yet completed the delete hence there is still a ContentTreeStatus

zedi-pramodh replied:
This code is executed during image creation so I did not understand why zedagent deleting the ContentTreeConfig is a valid case in here ?

eriknordmark commented on pkg/pillar/cas/containerd.go:
```diff
@@ -705,6 +708,85 @@ func (c *containerdCAS) PrepareContainerRootDir(rootPath, reference, rootBlobSha
 		logrus.Errorf(err.Error())
 		return err
 	}
+	// On kubevirt eve we also write the mountpoints and cmdline files to the rootPath.
+	// Once rootPath is populated with all necessary files, this rootPath directory will be
+	// converted to a PVC and passed in to domainmgr to attach to VM as rootdisk
+	// NOTE: For kvm eve these files are generated and passed to bootloader in pkg/pillar/containerd/oci.go:AddLoader()
```
Is the pillar/cas package the best place to do this for kubevirt? Or would it make more sense to do it in a structurally similar place e.g., in the pillarcontainerd package?

zedi-pramodh replied:
I believe pillar/cas is the best place for kubevirt eve. Because we need to have a rootfs directly laid out with all files before we convert to PVC. After that only PVC will be used in case of kubevirt eve. For kvm eve doing in pkg/pillar/containerd/oci.go:AddLoader() makes sense because it uses OCI spec.

eriknordmark commented on pkg/pillar/cas/containerd.go:
```diff
@@ -705,6 +708,85 @@ func (c *containerdCAS) PrepareContainerRootDir(rootPath, reference, rootBlobSha
+	if base.IsHVTypeKube() {
+
+		// Mount the snapshot
+		if err := c.MountSnapshot(snapshotID, GetRoofFsPath(rootPath)); err != nil {
```
You leave the mount in place once this is done.
Does that mean there is a contract with some code which runs later where it assumes it is already mounted? That might be a reason to move this code, or if not document this contract in comments.

zedi-pramodh replied:
Yes we need this mount to exist so that this container directory will be converted to qcow2->PVC. But the unmount happens after qcow2 creation in pkg/pillar/volumehandlers/csihandler.go.

Sure I will add  comments in here.

eriknordmark commented on pkg/pillar/cmd/volumemgr/updatestatus.go:
```diff
@@ -592,7 +608,16 @@ func doUpdateVol(ctx *volumemgrContext, status *types.VolumeStatus) (bool, bool)
+				cfg := lookupContentTreeConfig(ctx, status.ContentID.String())
+				if cfg != nil && cfg.IsAppImage && ctStatus.OciImageName != "" {
+					imgName = ctStatus.OciImageName
+					status.IsAppImage = true
```
When you modify status you need to set changed to true so that the function returns true (and the caller hence publishes the status).
The logic here makes it hard to tell whether you modified IsAppImage since you force it to false above.

eriknordmark commented on pkg/pillar/cmd/volumemgr/updatestatus.go:
```diff
@@ -614,7 +639,7 @@ func doUpdateVol(ctx *volumemgrContext, status *types.VolumeStatus) (bool, bool)
 				return changed, false
 			}
-			status.ReferenceName = ctStatus.ReferenceID()
+			status.ReferenceName = imgName
```
FWIW this is fine since changed is set below.

eriknordmark commented on pkg/pillar/types/contenttreetypes.go:
```diff
@@ -131,6 +132,8 @@ type ContentTreeStatus struct {
 	NameIsURL    bool
 	// Blobs the sha256 hashes of the blobs that are in this tree, the first of which always is the root
 	Blobs []string
+	// kube oci image
```
This comment doesn't tell me what it is.

zedi-pramodh replied:
Will add comments.

eriknordmark commented on pkg/pillar/types/contenttreetypes.go:
```diff
@@ -26,6 +26,7 @@ type ContentTreeConfig struct {
 	GenerationCounter int64
 	DisplayName       string
 	CustomMeta        string
+	IsAppImage        bool
```
I'm guessing this is to tell apart infrastucture pieces like the k3s, longhorn, etc pieces.
Or is it to tell apart app images and baseos images??
If the former it would make sense to pick a different name and explain in a comment.

zedi-pramodh replied:
This is for the eve app image. Nothing to do with the kubernetes components.

eriknordmark replied:
Why does the code in volumemgr now need to care about app images vs. baseos images? Where are the comments explaining this?

zedi-pramodh replied:
I will add comments. Basically since we will be supporting native container launch we need prepend the docker.io prefix to the images we are planning to launch as native containers.  The prefix is applied in volumegr/handlework.go:mayAppendImgPrefix(). But to apply the prefix we need to know if the image is an eve container app or not. We now keep that information in ContentTreeConfig.

eriknordmark replied:
Don't we have IsContainer() for this?

zedi-pramodh replied:
Yes good point. I am looking into that with Naiming and will update.

zedi-pramodh replied:
Fixed this code and removed the flag IsAppImage and used IsContainer() wherever needed. Submitted a new commit.

eriknordmark commented on pkg/pillar/types/volumetypes.go:
```diff
@@ -398,18 +420,30 @@ type VolumeRefStatus struct {
 	VerifyOnly             bool
 	Target                 zconfig.Target
 	CustomMeta             string
+	// kube pod container image
```
What is this? Comment doesn't explain.

eriknordmark replied:
Is is still there. But it wouldn't hurt to explain what ReferenceName is.

zedi-pramodh replied:
I will remove the comment but keep the ReferenceName.  We actually copy the VolumeStatus.ReferenceName to VolumeRefStatus.ReferenceName.  If you want to change then we need to change VolumeStatus too and all over the code.

eriknordmark commented on pkg/pillar/volumehandlers/csihandler.go:
```diff
@@ -0,0 +1,330 @@
+}
+
+func (handler *volumeHandlerCSI) CreateSnapshot() (interface{}, time.Time, error) {
+	//TODO implement me
```
Would make sense to create a Jira ticket so we don't forget this later.

zedi-pramodh replied:
Sure will do.

eriknordmark reviewed (approved):
Some suggestions/questions around code structure and some things to fix around '#' and generationCounter, etc.

eriknordmark reviewed (changes requested):
Some suggestions/questions around code structure and some things to fix around '#' and generationCounter, etc.

christoph-zededa commented:
> Thanks @christoph-zededa can I actually go ahead and use the code you suggested ?

Sure.

zedi-pramodh commented on pkg/pillar/types/volumetypes.go:
```diff
@@ -137,12 +142,19 @@ type VolumeStatus struct {
 func (status VolumeStatus) Key() string {
+
+	// PVC names should not include # so lets choose -pvc-
+	if base.IsHVTypeKube() {
```
Yes VolumeKey() and Key() are being used as filenames. Domainmgr, Zedagent, Zedmanager are using those keys (filenames). I just followed same convention not to break anything. IMO, changing that now should be separate effort independent of this commit. I am not sure what all we break (upgrades ?)

eriknordmark replied:
Are you talking about the fact that the keys in pubsub are used as filenames for the checkpoint files (in /run/<agentname> and /persist/status/<agentname>) or some other use?

My understanding is that the PVCs can not have names with '#' but domainmgr etc do not manipulate PVCs; they use the FileLocation field to determine what to pass to the hypervisor for mounts.

zedi-pramodh replied:
PathName() is used only during CreateVolume(), I did not add that for csihandler (Kubevirt) because PVC is in Kubernetes metadata and not a explicit file on disk like qcow2 or zvols.

eriknordmark replied:
Where do you end up using VolumeKey() and/or Key() to  name the PVC? That seems like a bug to me independent of your changes.

zedi-pramodh replied:
Yes I commented above on this question. Basically Domainmgr, Zedagent and Zedmanager are generating the filenames using VolumeKey() and Key(). Not sure if I can change that now as part of this PR.

eriknordmark replied:
There are only 18 instances of VolumeKey in pkg/pillar and I don't see how any of them get fed into the filename.
(It is a lot harder to look for the usage of Key() since the name is used for lots of structs.)
Hence I'm asking for specifics of what broke where.

eriknordmark replied:
Looking some more at volumemgr and domainmgfr (which are the ones involved in the volume creation and use) I do not see any place where Key() is used. Instead we have various fields (FileLocation, Target, Path, PathName) which track the file/pathname.

This seems to start with using func (status VolumeStatus) PathName() 

Thus I wonder if you've added some use of Key() or VolumeKey() in your branch which is causing issues.
FWIW it might make sense to introduce a new func (status VolumeStatus) VolumeFileName() which produces the filename part and which PathName can use.

Something like (derived from current master code - you'd add the IsHvType in VolumeFileName()):
`// PathName returns the path of the volume
func (status VolumeStatus) PathName() string {
	baseDir := VolumeClearDirName
	if status.Encrypted {
		baseDir = VolumeEncryptedDirName
	}
	return filepath.Join(baseDir, status.VolumeFileName())
}

// VolumeFileName returns the basename path of the volume
func (status VolumeStatus) VolumeFileName() string {
	return fmt.Sprintf("%s#%d.%s",
		status.VolumeID.String(),
		status.GenerationCounter+status.LocalGenerationCounter,
		strings.ToLower(status.ContentFormat.String()))
}

`

zedi-pramodh replied:
Key() is used to lookup the volumeconfig in LookupVolumeConfig()
Key() is used to lookup the VolumeRefConfig in lookupVolumeRefConfig()
Key() is used to lookup the VolumeStatus in LookupVolumeStatus()

Also obviously same key() is used to publish the corresponding VolumeStatus, VolumeConfig and VolumeRefConfig.

So, Key() is used AFAIK.

Now the question is do we really need to replace "#" with "-pvc-" for kubevirt to generate a Key() which is just used to index into lookup table ?? May be not, I guess that is what you are trying to point out. I will revert the changes to Key() and VolumeKey() and implement the PathName() as you suggested and see if the tests pass.

BTW, just in case you are wondering how everything is working, I made similar changes in zedagent, zedmanager etc where Key() is generated,  which are not part of this commit. So functionality wise nothing is broken but certainly looks unnecessary.

eriknordmark replied:
Time for a call.
Of course Key() is used - for pubsub. But never for file/volume names.
So not necessary.

eriknordmark replied:
Resolved in most recent code.

eriknordmark commented on pkg/pillar/cmd/volumemgr/handlework.go:
```diff
@@ -248,6 +253,13 @@ func casIngestWorker(ctxPtr interface{}, w worker.Work) worker.WorkResult {
 	if err != nil {
 		result.Error = err
 		result.ErrorTime = time.Now()
+	} else {
+		if ctx.hvTypeKube {
+			cfg := lookupContentTreeConfig(ctx, status.ContentID.String())
+			if cfg != nil && status.IsContainer() {
+				publishContentTreeStatus(ctx, &status)
+			}
```
Why do you need this code? Are you missing a publish somewhere else?

eriknordmark replied:
Or is this leftover code from debugging something else?

eriknordmark replied:
Still don't have an answer to this question.

zedi-pramodh replied:
Actually not needed I guess. This code was written many months ago to get the native container part working, I do not see any reason to publish content tree status from here. I will clean up.

eriknordmark reviewed (changes requested):
Your bug is in three occurances of
pvcName := handler.status.Key()
and one occurance of kubeapi.FindPVC(handler.status.Key()

Those should not use Key() but the VolumeFileName() function I suggested.

Also, you probably want to change the log statements which logs the PVC name in that function to use the same VolumeFileName().

zedi-pramodh commented:
Yes I found that yesterday when I fixed the VolumeKey and Key. Basically now added a new function GetPVCName() which generates the PVC name from status. I am testing the fix with ztest automation since yesterday.

zedi-pramodh commented:
Fixed the bug and updated this PR with a new commit

eriknordmark commented on Makefile:
```diff
@@ -392,7 +392,7 @@ currentversion:
 test: $(LINUXKIT) test-images-patches | $(DIST)
 	@echo Running tests on $(GOMODULE)
-	$(QUIET)$(DOCKER_GO) "gotestsum --jsonfile $(DOCKER_DIST)/results.json --junitfile $(DOCKER_DIST)/results.xml --raw-command -- go test -coverprofile=coverage.txt -covermode=atomic -race -json ./..." $(GOTREE) $(GOMODULE)
+	$(QUIET)$(DOCKER_GO) "gotestsum --jsonfile $(DOCKER_DIST)/results.json --junitfile $(DOCKER_DIST)/results.xml --raw-command -- go test -tags kubevirt -coverprofile=coverage.txt -covermode=atomic -race -json ./..." $(GOTREE) $(GOMODULE)
```
@christoph-zededa  can you review this test thing? I don't know what the tags do.

christoph-zededa replied:
Then I would review my own code: https://github.com/lf-edge/eve/pull/3768#issuecomment-1966813344 ;-)

It is passing the build flag for kubevirt to build the tests with this flag set.

eriknordmark replied:
:-) :-)
So the assumption is that with kubevirt virtualization mode we'll run a superset of the tests, which is presumably true now. 
Down the road we might have different subsets of tests for different variants, but we can cross that bridge when we get to it.

eriknordmark commented on pkg/pillar/types/volumetypes.go:
```diff
@@ -178,6 +168,13 @@ func (status VolumeStatus) PathName() string {
 }
 
+// PVCName returns the volume name for kubernetes(longhorn)
+func (status VolumeStatus) GetPVCName() string {
```
Someone will be reading this code in 5 years and wonder why.
So please add back your comment that these can not contain special characters like e.g., "#"

eriknordmark commented on pkg/pillar/cmd/volumemgr/handlevolume.go:
```diff
@@ -192,6 +193,26 @@ func publishVolumeStatus(ctx *volumemgrContext,
 	status *types.VolumeStatus) {
 
 	key := status.Key()
+	if ctx.hvTypeKube {
+		vrStatus := lookupVolumeRefStatus(ctx, key)
+		sub := ctx.pubContentTreeStatus
+		items := sub.GetAll()
+		reference := ""
+		for _, item := range items {
+			cts := item.(types.ContentTreeStatus)
+			if status.ContentID.String() == cts.ContentID.String() && cts.IsContainer() {
+				reference = cts.ReferenceID()
+				break
+			}
+		}
+		if vrStatus != nil && reference != "" {
+			if !strings.HasPrefix(vrStatus.ReferenceName, types.KubeContainerImagePrefix) && vrStatus.IsContainer() {
+				vrStatus.ReferenceName = types.KubeContainerImagePrefix + reference
+				log.Functionf("publishVolumeStatus: sync reference name %s", vrStatus.ReferenceName)
+				publishVolumeRefStatus(ctx, vrStatus)
+			}
+		}
+	}
```
Why do these extra publish belong here?
volumemgr has code which makes sure that a change to a VolumeStatus is reflected to all the VolumeRefStatus which depend on it. Is that code not working for kubevirt? Not being invoked?
Or is this leftover code which is not needed any more?

eriknordmark commented on pkg/pillar/cmd/volumemgr/initialvolumestatus.go:
```diff
@@ -117,6 +140,38 @@ func gcVolumes(ctx *volumemgrContext, locations []string) {
 	}
 }
 
+func getVolumeStatusByPVC(pvcName string) (*types.VolumeStatus, error) {
```
How about adding a comment that this is the reverse of the PVCName function you added? That would be helpful for a reader.

eriknordmark commented on pkg/pillar/cmd/volumemgr/updatestatus.go:
```diff
@@ -460,6 +464,13 @@ func doUpdateVol(ctx *volumemgrContext, status *types.VolumeStatus) (bool, bool)
+			ctStatus := ctx.LookupContentTreeStatus(status.ContentID.String())
+			if ctStatus.IsContainer() {
+				log.Functionf("doUpdateVol: reference name %v, return changed", status.ReferenceName)
+				return true, false
```
Why are you returning changed=true here? This function didn't modify the VolumeStatus hence no need to do that.

zedi-pramodh replied:
Yeah removed that.

eriknordmark commented on pkg/pillar/types/volumetypes.go:
```diff
@@ -143,6 +143,7 @@ type VolumeStatus struct {
 
 // Key is volume UUID which will be unique
 func (status VolumeStatus) Key() string {
+
```
Accidentally added blank line?

zedi-pramodh replied:
Done

eriknordmark commented on pkg/pillar/cmd/volumemgr/handlework.go:
```diff
@@ -235,8 +235,13 @@ func casIngestWorker(ctxPtr interface{}, w worker.Work) worker.WorkResult {
 		}
 	}
 
+	appImgName := status.ReferenceID()
+	if ctx.hvTypeKube && status.IsContainer() {
+		appImgName = types.KubeContainerImagePrefix + appImgName
+	}
```
You've changed all but two of the callers to ReferenceID() to add this prefix, but  the conditions are not identical - some check for additional things. I don't know if that is intentional or accidental.

Wouldn't it make sense to instead change ReferenceID() to do this in one place?
If so then ctx.hvTypeKube needs to be either an argument to ReferenceID, or set as a new boolean in the ContentTreeStatus when it is created.
That level of uniformity/consistency and information hiding would make this easier to understand and maintain.
2)

eriknordmark reviewed (changes requested):
Thanks for taking care of the Key and VolumeKey.

I've resolved the issues which I see being addressed in the code.
Some additional comment nits, 
- I think the ReferenceID (and misnamed ReferenceName) needs to be made consistent and not sprinked slightly differently in almost all callers
- the update of ReferenceID in VolumeRefStatus and associated publish is in the wro ng place

zedi-pramodh commented:
Oh I see what you are saying. Looks like we are Publishing VolumerefStatus in VolumeStatus. Need to check with @naiming-zededa I believe he added that code long time ago. I will rework that and see if tests pass.

eriknordmark commented:
> Oh I see what you are saying. Looks like we are Publishing VolumerefStatus in VolumeStatus. Need to check with @naiming-zededa I believe he added that code long time ago. I will rework that and see if tests pass.

Code has added passing the ReferenceID back to zedmanager by adding it to VolumeRefStatus. I assume this is so zedmanager can pass that on to domainmgr/kubevirt. So that is fine. But the implementation has code sprinkled in somewhat random places which makes it hard to understand and maintain.

eriknordmark commented:
BTW: you have some conflicts to resolve before the next time we can get the PR to build and run Eden

zedi-pramodh commented:
Cleaned up the unnecessary code and resolved merge conflicts with master. Published new commit

eriknordmark commented on pkg/pillar/types/contenttreetypes.go:
```diff
@@ -157,6 +160,10 @@ func (status ContentTreeStatus) IsContainer() bool {
 // ReferenceID get the image reference ID
 func (status ContentTreeStatus) ReferenceID() string {
+
+	if base.IsHVTypeKube() && status.IsContainer() {
```
base.IsHVTypeKube opens and reads a file, right?
I don't think we want to do that all the time.
So either pass that as an argument or make it be set in a field in ContentTreeStatus.

zedi-pramodh replied:
Yeah I will add to ContentTreeStatus

zedi-pramodh replied:
Done

eriknordmark commented on pkg/pillar/cas/containerd_test.go:
```diff
@@ -0,0 +1,43 @@
+package cas
```
This file is missing Copyright and license lines.

zedi-pramodh replied:
Done

milan-zededa commented on pkg/pillar/Makefile:
```diff
@@ -94,7 +94,7 @@ test: build-docker-test
 		--jsonfile /pillar/results.json \
 		--junitfile /pillar/results.xml \
-		--raw-command -- go test -coverprofile=coverage.txt -covermode=atomic -race -json \
+		--raw-command -- go test -tags kubevirt -coverprofile=coverage.txt -covermode=atomic -race -json \
```
Is there a unit test that failing to build without the kubevirt tag?
If so, we should put it under `//go:build kubevirt`

zedi-pramodh replied:
Not that I know of. This code is given by @christoph-zededa I just used it.

christoph-zededa replied:
There is no unit test that is failing without the kubevirt tag.

There are two reasons I've put it there:
* someone might add test for code that is under the `kubevirt` build flag
* make sure that the current tests are still successful even with the kubevirt build flag set

eriknordmark reviewed (approved):
I still don't understand why the same field is called ReferenceID in some structure ReferenceName in others (plus local variables which contain the same are called things like imgname (I think).

Kick off tests.
Please squash the commits. I think one commit makes sense in this case but your call as long as each commit in the patch set can stand on its own. (Sometimes the tests and docs are in separate commits.)

zedi-pramodh commented:
Yes those commits can be independent since I only cleaned up stuff and functionality is still the same.

eriknordmark commented:
> Yes those commits can be independent since I only cleaned up stuff and functionality is still the same.

You mean you want all 6 to be independent? How will that make sense to somebody else (a reader) a year from now or even a month from now? Please squash.

zedi-pramodh commented:
There is a rebase of master too in one of the commits to fix the merge conflicts. So not sure what will happen if I squash the commits.  I will try and see in worst case I will submit another PR pointing to this PR

eriknordmark commented:
If you have a merge commit in your PR you MUST squash it out. There is no problem doing that - git rebase -i upstream/master works for me.

zedi-pramodh commented:
Ok squash worked cleanly. Updated the single squashed commit.
